toolchains/symbiflow.py

```diff
@@ -1001,6 +1003,451 @@ def check_env():
+                        clocks[clk_name]['actual'] = float(
+                            "{:.3f}".format(clk_freq)
+                        )
+                        clocks[clk_name]['requested'] = float(
+                            "{:.3f}".format(req_clk_freq)
+                        )
+                        clocks[clk_name]['met'] = match.groups()[2] == "PASS"
+                        clocks[clk_name]['setup_violation'] = 0
+                        clocks[clk_name]['hold_violation'] = 0
+
+        return clocks
+
+    def get_resources(self):
+        """Returns a dictionary with the resources parsed from the nextpnr log file"""
+
+        resources = dict()
+
+        log_file = os.path.join(self.out_dir, 'nextpnr.log')
+
+        with open(log_file, "r") as file:
+            processing = False
+            for line in file:
+                line = line.strip()
+
+                if "Device utilisation" in line:
+                    processing = True
+                    continue
+
+                if not processing:
+                    continue
+                else:
+                    if len(line) == 0:
+                        break
+
+                    res = line.split(":")
+                    res_type = res[1].strip()
+
+                    regex = "([0-9]*)\/.*"
+                    match = re.match(regex, res[2].lstrip())
+
+                    assert match
+
+                    res_count = int(match.groups()[0])
+                    resources[res_type] = res_count
+
+        return resources
+
+    def resources(self):
+        lut = 0
+        dff = 0
+        carry = 0
+        iob = 0
+        pll = 0
+        bram = 0
+
+        res = self.get_resources()
+
+        if 'SLICE_LUTX' in res:
+            lut = res['SLICE_LUTX']
+        if 'SLICE_FFX' in res:
+            dff = dff = res['SLICE_FFX']
+        if 'CARRY4' in res:
+            carry = res['CARRY4']
+        if 'IOB33M_OUTBUF' in res:
+            assert res['IOB33M_OUTBUF'] == res['IOB33S_OUTBUF']
+            iob = iob + res['IOB33M_OUTBUF']
+        if 'IOB33M_INBUF_EN' in res:
+            #TODO: check if this assert is correct becasue for baselitex-nexys-video it fails
+            #assert res['IOB33M_INBUF_EN'] == res['IOB33S_INBUF_EN']
+            iob = iob + res['IOB33M_INBUF_EN']
+        if 'IOB33_OUTBUF' in res:
+            iob = iob + res['IOB33_OUTBUF']
+        if 'IOB33_INBUF_EN' in res:
+            iob = iob + res['IOB33_INBUF_EN']
+        if 'RAMB18E1_RAMB18E1' in res:
+            bram = res['RAMB18E1_RAMB18E1']
+        if 'RAMB36E1_RAMB36E1' in res:
+            bram += res['RAMB36E1_RAMB36E1'] * 2
+        if 'PLLE2_ADV_PLLE2_ADV' in res:
+            pll = res['PLLE2_ADV_PLLE2_ADV']
+
+        ret = {
+            "LUT": lut,
+            "DFF": dff,
+            "BRAM": bram,
+            "CARRY": carry,
+            "GLB": None,
+            "PLL": pll,
+            "IOB": iob,
+        }
+        return ret
```

Comment:
We need to think on how to deal with resources in a more generic way, as the FPGA interchange is meant to work for any kind of architecture and FPGA family and not be bound to one of them only (e.g. xc7).

This implementation of the resources function assumes that we are dealing with and xc7 device, which may not be true as soon as other architectures are available.

Reply:
I think we can introduce the [resources_map](https://github.com/SymbiFlow/fpga-tool-perf/pull/323/files#diff-69ab290f878020ed67b1008338e82df43674e16b1e45733d5f0d8e04b6601141R585-R604) in the generic part where we could add arch and vendor dependent keywords related to resources.